### ISSUE TITLE
Glitch in multi-channel backport

### ISSUE THREAD

**Issue by agraef** [pd-l2ork/pd-l2ork#58]
@MovementGH, this is wrong:
https://github.com/pd-l2ork/pd-l2ork/blob/5f39a7652345df806631253aa0da1dfe546d64ea/pd/src/d_osc.c#L222

I noticed this while pulling your changes over to purr-data.

cos_cleanup is *not* to be called as an object destructor (in pd_free), but as a class destructor (set with class_setfreefn, which we don't have in pd-l2ork/purr-data). Otherwise the cos tables get freed the first time a cos~ object gets freed, which causes subsequent crashes because the cos tables don't exist anymore. Just leave the freemethod in class_new as 0 there (so that cos_cleanup is *never* called), then everything will be fine.

Greetings from purr-data, and thanks for backporting all this stuff, that made my live a lot easier! :)

Albert

P.S.: I'm still working on a crash bug in signal_makereusable() which I suspect may be caused by a similar backporting glitch, but I haven't found the cause of that yet.

**Comment by MovementGH** [comment]
Thank you for letting me know, will address. Let me know if the signal_makereusable() bug ends up being related to the backports as well. There were quite a few issues to work through while backporting, it's more than likely that there are more that I missed (like possibly that one)

**Comment by agraef** [comment]
Yes, sure. I'm pretty sure that it's related to the backport, because it's the moonlib dinlet~ object which causes the crash. Apparently this is an old variant of inlet~ (the old version without the message forwarding and extra control in- and outlet) with an added float argument.

It uses an outdated t_vinlet declaration and initialization, so that's probably the cause for the data inconsistency that leads to signal_makereusable crashing. That by itself should be easy to fix. If I'm not mistaken, in pd-l2ork you still use the old (0.47?) inlet~ with just one signal outlet, so you should be able to make it work.

In purr-data, however, we already have the new message-forwarding inlet~, so I'm not sure whether I can make dinlet~ compatible with that. In that case it's probably best to just remove it.

**Comment by agraef** [comment]
Actually, I think that pd-l2ork might be unaffected by the crash in signal_makereusable, because you still use an older version of g_io.c, so dinlet~'s copy of of t_vinlet could still be compatible. But I'd check it anyway.

**Comment by agraef** [comment]
I can confirm that fixing up data structure and initialization of moonlib/dinlet~ fixes the crash in purr-data. Whether you need similar changes I'm not sure; if `make check` runs without crashing then pd-l2ork should be fine. (You'd first have to fix up your version of the regression tests, though, because `make check` was broken in pd-l2ork last time I tried it.)

**Comment by MovementGH** [comment]
Great, thanks. The regression tests do not crash, so we might be unaffected as you have said. That said, thanks for mentioning the regression tests, I was unaware of their existence, so they indeed have not been updated with some of the newer backports (for example, as a part of the backports I have been working on, I backported the inlet~ message forwarding).

**Comment by pd-l2ork** [comment]
Unless there are any other extant issues, i will be closing this one since there are 2 fixes addressing the main point here. Do we want to also port the t_vinlet? @MovementGH has this been done since?

**Comment by pd-l2ork** [comment]
Does the lack of vinlet port break out functional (not code) compatibility with vanilla?
